### PR TITLE
Rework DiscordJS.Client access

### PR DESCRIPTION
**Pull request by Blatoy**
Progress for #7, we still need to catch the errors in the API and at other places for it to be fixed now.

This will also make it easier to write unit tests in the future.

### CODE REVIEW

File: bot/apis/discord-client-provider.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const { Client } = require("discord.js");
+
+class DiscordAPIConnectionError extends Error {
+    constructor(message) {
+        super(message);
+        this.name = "DiscordAPIConnectionError";
+    }
+}
+
+class DiscordClientProvider {
```
Comment (RedMser) [inline bot/apis/discord-client-provider.js:20]: Add JS Docs to both classes

File: bot/index.js [hunk 2]
```diff
@@ -65,7 +65,7 @@ async function init() {
         }
     }
 
-    global.discordClient._connectToDiscord = connectToDiscord;
+    global.discotron._connectToDiscord = connectToDiscord;
```
Comment (RedMser) [inline bot/index.js:67]: Should it be renamed to `connectToDiscord` without the `_`, now that it is a global object that we have full control over its properties?

File: bot/classes/user-role.js [hunk 3]
```diff
@@ -42,8 +42,9 @@ class UserRole extends UserRoleModel {
         if (this.discordUserId !== null) {
             return this.discordUserId === userDiscordId;
         } else {
-            if (typeof global.discordClient !== "undefined" && typeof global.discordClient.guilds.get(this.discordGuildId) !== "undefined") {
-                let role = global.discordClient.guilds.get(this.discordGuildId).roles.get(this.discordRoleId);
+            const discordClient = discordClientProvider.get();
+            if (typeof discordClient !== "undefined" && typeof discordClient.guilds.get(this.discordGuildId) !== "undefined") {
```
Comment (RedMser) [inline bot/classes/user-role.js]: This undefined check is no longer needed, since `.get()` throws when we are offline.

**Review by RedMser — CHANGES_REQUESTED** [review]

File: bot/index.js (hunk 2 above)
Comment (Blatoy) [reply bot/index.js:67]: I think we should leave it, because this should disappear anyway in the future.

File: bot/classes/user-role.js [hunk 4]
```diff
@@ -42,11 +42,12 @@ class UserRole extends UserRoleModel {
         if (this.discordUserId !== null) {
             return this.discordUserId === userDiscordId;
         } else {
-            const discordClient = discordClientProvider.get();
-            if (typeof discordClient !== "undefined" && typeof discordClient.guilds.get(this.discordGuildId) !== "undefined") {
-                let role = discordClient.guilds.get(this.discordGuildId).roles.get(this.discordRoleId);
+            try {
+                const discordClient = discordClientProvider.get();
+                const role = discordClient.guilds.get(this.discordGuildId).roles.get(this.discordRoleId);
```
Comment (RedMser) [inline bot/classes/user-role.js]: Inline the discordClient variable?

**Review by RedMser — CHANGES_REQUESTED** [review]

**Review by RedMser — APPROVED** [review]

**Pull request merged**